auth/validator.go

```diff
@@ -74,31 +76,37 @@ func getHeaders(encodedToken string) (jws.Headers, error) {
 	return msg.Signatures()[0].ProtectedHeaders(), nil
 }
 
-func (m *Middleware) validateClaims(t Token, ks *oidcclient.OIDCTenant) error { // performing IsExpired check, because dgriljalva jwt.Validate() doesn't fail on missing 'exp' claim
-	// performing IsExpired check, because lestrrat-go jwt.Validate() doesn't fail on missing 'exp' claim
-	if t.IsExpired() {
-		return fmt.Errorf("token is expired, exp: %v", t.Expiration())
+func (m *Middleware) validateClaims(t Token, ks *oidcclient.OIDCTenant) error { // performing isExpired check, because dgriljalva jwt.Validate() doesn't fail on missing 'exp' claim
+	// performing isExpired check, because lestrrat-go jwt.Validate() doesn't fail on missing 'exp' claim
+	if t.isExpired() {
+		return fmt.Errorf("token is expired, exp: %v", t.expiration())
 	}
 	err := jwt.Validate(t.getJwtToken(),
 		jwt.WithAudience(m.oAuthConfig.GetClientID()),
 		jwt.WithIssuer(ks.ProviderJSON.Issuer),
-		jwt.WithAcceptableSkew(1*time.Minute)) // to keep leeway in sync with Token.IsExpired
+		jwt.WithAcceptableSkew(1*time.Minute)) // to keep leeway in sync with Token.isExpired
 
 	if err != nil {
 		return fmt.Errorf("claim validation failed: %v", err)
 	}
 	return nil
 }
 
-func (m *Middleware) getOIDCTenant(tokenIssuer string) (*oidcclient.OIDCTenant, error) {
-	issURI, err := m.verifyIssuer(tokenIssuer)
+// getOIDCTenant returns an OIDC Tenant with discovered .well-known/openid-configuration.
+//
+// tokenIssuer is the iss claim of the incoming token
+//
+// iasIssuer is the ias_iss claim of the incoming token
+func (m *Middleware) getOIDCTenant(tokenIssuer, iasIssuer string) (*oidcclient.OIDCTenant, error) {
+	issURI, err := m.verifyIssuer(iasIssuer)
 	if err != nil {
 		return nil, err
 	}
 
-	oidcTenant, exp, found := m.oidcTenants.GetWithExpiration(tokenIssuer)
-	if !found || time.Now().After(exp) {
-		newKeySet, err, _ := m.sf.Do(tokenIssuer, func() (i interface{}, err error) {
+	oidcTenant, exp, found := m.oidcTenants.GetWithExpiration(iasIssuer)
+	// redo discovery if not found, cache expired, or tokenIssuer is not the same as Issuer on providerJSON (e.g. custom domain config just changed for that tenant)
+	if !found || time.Now().After(exp) || oidcTenant.(*oidcclient.OIDCTenant).ProviderJSON.Issuer != tokenIssuer {
```

Comment:
when i'm right, the caching here is the only reason, why you need the "issuer" method, correct?